assets/app/scripts/services/data.js

```diff
@@ -514,55 +521,55 @@ DataService.prototype.createStream = function(resource, name, context, opts, isR
 //                    a websocket. Default is false.
 //            pollInterval: in milliseconds, how long to wait between polling the server
 //                    only applies if poll=true.  Default is 5000.
-//
+//            http:   similar to .get, etc. at this point, only used to pass http.params for filtering
 // returns handle to the watch, needed to unwatch e.g.
 //        var handle = DataService.watch(resource,context,callback[,opts])
 //        DataService.unwatch(handle)
   DataService.prototype.watch = function(resource, context, callback, opts) {
     resource = APIService.toResourceGroupVersion(resource);
     opts = opts || {};
-
+    var key = this._uniqueKeyForResourceContext(resource, null, context, _.get(opts, 'http.params'));
     if (callback) {
       // If we were given a callback, add it
-      this._watchCallbacks(resource, context).add(callback);
+      this._watchCallbacks(key).add(callback);
     }
-    else if (!this._watchCallbacks(resource, context).has()) {
+    else if (!this._watchCallbacks(key).has()) {
       // We can be called with no callback in order to re-run a list/watch sequence for existing callbacks
       // If there are no existing callbacks, return
       return {};
     }
 
-    var existingWatchOpts = this._watchOptions(resource, context);
+    var existingWatchOpts = this._watchOptions(key);
     if (existingWatchOpts) {
       // Check any options for compatibility with existing watch
       if (existingWatchOpts.poll != opts.poll) {
         throw "A watch already exists for " + resource + " with a different polling option.";
       }
     }
     else {
-      this._watchOptions(resource, context, opts);
+      this._watchOptions(key, opts);
     }
 
     var self = this;
 
-    if (this._watchInFlight(resource, context) && this._resourceVersion(resource, context)) {
+    if (this._watchInFlight(key) && this._resourceVersion(key)) {
       if (callback) {
         $timeout(function() {
-          callback(self._data(resource, context));
+          callback(self._data(key));
```

Comment:
agreed, thats better

On Thu, Apr 28, 2016 at 2:49 PM, Ben Petersen <notifications@github.com>
wrote:

> In assets/app/scripts/services/data.js
> <https://github.com/jwforres/origin/pull/40#discussion_r61482085>:
>
> >        if (callback) {
> >          $timeout(function() {
> > -          callback(self._data(resource, context));
> > +          callback(self._data(key));
>
> Inclined to update it to this:
>
> // (!!this._data(key)) tells us here if there is something in the cache...if (this._watchInFlight(key) && this._resourceVersion(key) && (!!this._data(key))) {
>       if (callback) {
>         $timeout(function() {
>           callback(self._data(key));
>         }, 0);
>       }
>     }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jwforres/origin/pull/40/files/2ca135eab32bc6d2467d9276ff936b136179463e#r61482085>
>